second addition of module API.

shahzebsiddiqui opened the pull request:
replace subprocess.Popen with BuildTestCommand
raise TypeError in Module and ModuleCollection class when invalid
arguments passed for module or collection name.
Add method save(), describe() that implement saving and showing module
collections.
Change syntax of module load command to use command chaining (&&) which
is important when loading multiple modules and testing for return code.
Add docstring for all methods.
Add one example python to demonstrate how to save and show module
collections.

shahzebsiddiqui committed:
second addition of module API.
replace subprocess.Popen with BuildTestCommand
raise TypeError in Module and ModuleCollection class when invalid
arguments passed for module or collection name.
Add method save(), describe() that implement saving and showing module
collections.
Change syntax of module load command to use command chaining (&&) which
is important when loading multiple modules and testing for return code.
Add docstring for all methods.
Add one example python to demonstrate how to save and show module
collections.

vsoch commented on buildtest/modules/util.py:
```diff
@@ -1,48 +1,156 @@
-    )
-    out = ret.communicate()[1]
-    tmp = out.decode("utf-8").split()
+    cmd = BuildTestCommand()
```
I'm concerned that here we have shell set to True, which I think is considered bad practice:

```bash
        self.ret = subprocess.Popen(
            cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )
```
I think there are other ways to handle this (e.g., for sregistry I use Popen, see [here](https://github.com/singularityhub/sregistry-cli/blob/60e2e48d0fbf650f459d99f7b1d28f0c1e44069d/sregistry/utils/terminal.py#L126). Overall, having a consistent function to execute commands is great - but I don't think we want to use shell=True. I'll open an issue to refactor this for a future PR.

vsoch replied:
https://github.com/HPC-buildtest/buildtest-framework/issues/182

vsoch commented on buildtest/modules/util.py:
```diff
@@ -1,48 +1,156 @@
+    cmd = BuildTestCommand()
+    cmd.execute(collections)
+    # The output is captured in STDERR stream.
+    out = cmd.get_error().split()
```
I very much like the design of this command object - that we can execute, get_error, I'd even add another function to make it easy to print (preview) what command is going to be issued - this would be helpful to hook up to a debug mode.

shahzebsiddiqui replied:
let me try to add a method to ``BuildTestCommand()`` to get that information.

vsoch commented on buildtest/modules/util.py:
```diff
@@ -1,48 +1,156 @@
+    """Class declaration of ModuleCollection."""
+
     def __init__(self, collection):
+        """Initializer method of ModuleCollection class.
```
Can you walk me through the difference in just a module / module collection? I'm guessing it's an organized group of modules - is this something created by buildtest, or associated with Lmod? How does this link up to a user request to run just one recipe file:

```bash
buildtest build -c /path/to/mpi.yml
```
Versus an entire folder?

```bash
buildtest build -d /path/to/collection
```
Couldn't it be the case that there is a recipe (defined in a collection) that you would target the individual file for? Or maybe what I'm asking - what's the distinction between a collection and just a folder / subset of recipes, and given that we can organize with folders, why do we need collections?

shahzebsiddiqui replied:
modules and module collection are means to load the modules which are going to be used when defining the YAML schema. Currently the buildtest allowed few ways to load modules, two of them are the module collection and buildtest own module collection. 

See https://buildtest.readthedocs.io/en/latest/build_subcommand/module_collections.html

The schema needs to be changed, before user must specify ``buildtest build -c <config> -co <collection>`` or ``buildtest build -c <config> -mc <buildtest-collection-ID>`` to allow loading modules. If none are specified then active modules are loaded.

The idea was having a Module API which would allow a mechanism to interact with module system. 

If you are new to Lmod user collection check out https://lmod.readthedocs.io/en/latest/010_user.html#user-collections. 

Idea is to allow buildtest to write tests where user can specify how to load modules. Either user can create a Lmod user collection. Or use the buildtest module collection see https://buildtest.readthedocs.io/en/latest/module_operation.html#module-collection-operation-buildtest-module-collection which was implemented to allow an arbitrary size of module collections. The idea was to use collection ID instead of unique names since user who use collection names need to write unique names and it will cause issue when there are alot of user collection.

The ``ModuleCollection`` class is provided to interact with module collection. The ``Module`` class is the main class that user can interact with modules.


I have few examples of the YAML schema in this wiki https://github.com/HPC-buildtest/buildtest-framework/wiki/Yaml-Schema-Draft that i was thinkin about the new schema. This is not implemented.

Hope that answers few questions and maybe leads you to more questions :)

vsoch replied:
Interesting - my intuition is that the typical sysadmin would store module "collections" on the filesystem in folders, so this seems like a feature that would be afforded by just having folders instead, and then possibly using an include / exclude filter. 

But actually, since it's integrated with Lmod, this is probably something that is used and should be represented here. I can get a better sense when I am able to create my own little collection, and set of modules! I'll also ask my group about how they use collections. Either way, I think we should be careful with the UI, because there are now several different ways to select some subset of modules (and too many might be confusing for the UI):

 - select a config directly with -c / --config (collections not relevant)
 - select a directory with -d (with some include/exclude filters) - are collections relevant here? They probably shouldn't be.
 - select one or more collections
 - don't select anything, but discover all defined in the config.

Anyway, I am +1 for having collections given that they are a big part of Lmod - I'll need to refresh my learning about how they work to best discuss the implementation.

shahzebsiddiqui replied:
Lmod collection has been around for quite some time. Load a bunch of modules and save them into a collection, then purge your modules and you can restore all your modules using ``module restore <collection>``. 

I think we may be confusing some terminology, the ``--config`` and supposedly ``-d`` (not implemented) are options to deal with which YAML file to process. Inside the YAML you can specify key and one of them could be how to load modules. 

In the backend, buildtest needs to call the Module API to load the appropriate module. 

For example this example is showing that we can use a collection name ``GCC`` to restore the module. Assuming that user has this collection saved. The collections are saved in $HOME/.lmod.d/ 

```
# module restore example for C
hello_ex1:
  type: singlesource
  description: "C++ hello world example"
  compiler: gnu
  source: hello.cpp
  scheduler: local
  # module collection to restore
  moduleload:
    collection:
      name:
       - GCC
  cxxflags:
    - "-O2"
```

There can be other examples such as ones below

```
# module load example with one version
hello_ex1:
  type: singlesource
  description: "C hello world example1"
  compiler: gnu
  source: hello.c
  scheduler: local
  moduleload:
    key:
      name: GCC
      version: ["6.4.0"]
  cflags:
    - "-O1"

# single test with all version
hello_ex2:
  type: singlesource
  description: "C hello world example2"
  compiler: gnu
  source: hello.c
  scheduler: local
  moduleload:
    key:
      name: GCC
      version: ["*"]
  cflags:
    - "-O1"
```


The second example where ``version: ["*"]`` is doing a test against all GCC modules. This is not implemented in the ModuleAPI and it is pretty simple to get this since I have most of the methods implemented already.

vsoch commented on buildtest/modules/util.py:
```diff
@@ -1,48 +1,156 @@
-        )
-        ret.communicate()
-        return ret.returncode
+        """Test the module collection by running ``module restore <collection>``.
```
I'm wondering if the ModuleCollection should have similar functions to Module, just of course for the collection? Imagine if we had a list, with a Module and a ModuleCollection:

```python
modules = [Module(), ModuleCollection()]
```

With common methods, we could loop through:

```python
for moduleset in modules:
    moduleset.test()
```
But as it currently is, we'd have to do:

```bash
for moduleset in modules:
    if isinstance(moduleset, ModuleCollection):
        moduleset.testt_collection()
    elif isinstance(moduleset, Module):
        moduleset.test_module()
```
Also, if a module collection truly is a collection of modules, why isn't the Module subclass part of the ModuleCollection? For example, let's say I can create a ModuleCollection and add modules based on a uniqueid:

```
collection = ModuleCollection()
moduleExample = Module(...)
collection["example"] = moduleExample
```
And then the module collection has an internal dict that can be looped over / interacted with in various ways to test.

```bash
def test_collection(self):
    for name, module in self.modules.items():
        print("Testing %s" % name)
        module.test_module()
```
And you could even try subclassing a dict if you wanted the ModuleCollection to act like one. Or have it be a variable like in the example above.

shahzebsiddiqui replied:
The ``ModuleCollection()`` at the moment, was designed to take a collection name so that we can retrieve the command ``module restore <collection>`` this command needs to be in the test script  to load a module collection when buildtest generates a script. 

When i was thinking about module specification in YAML file was to avoid hard coding the module logic in tests as it gets too confusing and complex for framework to handle. So i thought let Lmod do most of the work. 

If ``ModuleCollection()`` is a subclass of ``Module()`` I am not sure how user will interact with this class. 

The ``Module().save()`` and ``Module().describe()`` are two methods that save and show module collection. It is assumed that user can specify which modules to load in the class that get saved. The ``ModuleCollection()`` is assumed to work only with collection that are saved. For instance listing all module collection is just one part of it.

vsoch replied:
I don't mean that ModuleCollection should be a subclass, but rather that ModuleCollection should ultimately load Module objects into it. It's exactly what it sounds like - a collection of actual Modules. This way, when the Module function is developed more, all of that functionality will be available with the ModuleCollection (because it includes Modules somewhere).

shahzebsiddiqui replied:
I'd be open to refactor ModuleCollection class in subsequent PRs. The PR is kinda big and complex. Could we do this in subsequent PRs.

vsoch replied:
Definitely!

vsoch commented on buildtest/modules/util.py:
```diff
@@ -1,48 +1,156 @@
+        """ Get the module command used to restore a collection
+
+        :return: Return the actual command to restore a collection.
+        :rtype: str
```
What is an example of this command? I'm not super familiar with Lmod / collections.

shahzebsiddiqui replied:
See https://github.com/shahzebsiddiqui/buildtest-framework/blob/module_api_v2.0/examples/collection.py#L14 this will return the module collection command. This is useful during test creation. Look at https://buildtest.readthedocs.io/en/latest/build_subcommand/module_collections.html#building-test-with-lmod-user-collection-buildtest-build-collection this example shows a CUDA collection loaded in test script to test a saxpy code. This is a way for users to save module collection and specify them in YAML file without having to worry about actual module names. 

Once you read https://lmod.readthedocs.io/en/latest/010_user.html#user-collections you will get idea why user collection are important.

vsoch commented on examples/saving_collection.py:
```diff
@@ -0,0 +1,17 @@
+import os
+import sys
+
+sys.path.insert(0, os.getenv("BUILDTEST_ROOT"))
```
We probably don't need this envar anymore, if it's installed on the Python path in site-packages.

vsoch commented on examples/saving_collection.py:
```diff
@@ -0,0 +1,17 @@
+
+from buildtest.modules.util import Module
+
+a = Module(["zlib"])
```
Why is this required to be a list? Intuitively, I'd just provide a string.

shahzebsiddiqui replied:
My thought was around dealing with module load command

```
self.module_load_cmd = [f"module load {x} && " for x in self.modules]
```

If user specify 
```
a = Module(["zlib","GCCcore","CUDA/10.0"]
```

This should generate a module load command as follows:

```
module load zlib && module load GCCcore && module load CUDA/10.0
```

It was easier to process a list since i need to run module load against all modules. Note that running ``module load zlib GCCcore CUDA/10.0`` is also valid but it does not guarantee all the modules are loaded. It is important when it comes to checking return code of the module load command.

You can try this in your system and see output of ``echo $?``  with command chaining like how i propose version doing ``module load <a> <b> <c>``.

Initialliy i had done ``module load <a>; module load <b>; module load <c>`` and this didnt work because it did not guarantee all modules were loaded since exit code will be of the last command.

Also in order for me to implement the ``Module().save()`` method i needed modules to be loaded in order to save them.

So running ``module load zlib && module save zlib`` worked when passing this to ``BuildTestCommand``

vsoch replied:
Sorry, what I mean is to allow:

```python
a = Module("zlib")
```
And then in the init, do something like:

```python
if not isinstance(name, list):
    name = [name]
```
That way, both of these will work:

```python
a = Module("zlib")
b = Module(["zlib", "mpi"])
```
Because it's not super intuitive to provide a list with one item to a function instantiation. The internal stuffs that you detailed looks good to me :)

shahzebsiddiqui replied:
got it. Let me fix this i think i get what you mean.

vsoch commented on examples/saving_collection.py:
```diff
@@ -0,0 +1,17 @@
+# save module collection with name "zlib"
+a.save("zlib")
+# show module collection "zlib"
+a.describe("zlib")
```
I think what we need (actually, what I need) is a very small bash script example that shows me how to set up a small local site with my own modules. And then provide instructions to run these commands. This would give me the entire picture:

 - being on a cluster with Lmod
 - creating a dummy module
 - interacting with buildtest to test it

Without any of the bullets above, a new user like myself won't have the full picture.

vsoch reviewed (commented):
Okay - so my review is only going to be superficial, because I don't think we are at a stage where things are properly working as we want, and I haven't experienced a build yet. So I'll review superficially, and do most of the more detailed testing with next step changes.

shahzebsiddiqui committed:
add debug support in Module() and ModuleCollection()
allow string to be passed in to Module() class.
add few more examples for testing purposes.

vsoch commented on buildtest/modules/util.py:
```diff
@@ -59,24 +70,26 @@ def get_command(self):
         if not isinstance(modules, list):
-            raise TypeError(f"Type Error: {collection} is not of type list")
+            self.modules = [modules]
 
```
I think you would still need to do:

```
if not isinstance(modules, list):
      modules = [modules]
self.modules = modules
```
No matter what is input (list or string) you want self.modules set.

shahzebsiddiqui replied:
i tried what you did, and it turns out this breaks badly. When running ``test_modules()`` you get the following output which is not right. For now it needs to be before, assumption was if its a list it is fine, otherwise it converts to a list. This logic can be revisited. I personally like throwing the ``TypeError()`` to be very explicit to on what type of argument type can be passed into `Modules` class.

```
[DEBUG] Executing module command: module purge && module load B &&  module load i &&  module load s &&  module load o &&  module load n &&  module load / &&  module load 3 &&  module load . &&  module load 0 &&  module load . &&  module load 4
```

vsoch replied:
This suggests that you are still looping through a string (the module name) and not the list - are you sure about your implementation?

vsoch commented on buildtest/modules/util.py:
```diff
@@ -149,8 +177,14 @@ def describe(self, collection="default"):
         out = cmd.get_error()
 
+        # print executed command for debugging
+        if self.debug:
```
This is a lot of redundant code - can this be a shared function for the class? I still think there could be a very basic base class for ModuleCollection and Module, or at least a shared function added to both to essentially run a command and print debug output (or not).

shahzebsiddiqui replied:
I think it makes sense to have a shared method for debug purpose.

It will require some code refactor and changes to example program. 

Id be open to changing the ModuleCollection class. I think this PR is already big enough.

I'd eventually like to have this part of regression test. In subsequent PRs.

Also Travis builds a mini eb stack for testing some of these features. That's a whole separate discussion

vsoch replied:
For this kind of debugging I usually put the logger in charge of this - you import the logger and then can just do logger.debug / logger.info and it will show (or not) depending on the user selected level.

vsoch commented on examples/saving_collection.py:
```diff
@@ -1,11 +1,11 @@
 import sys
 
-sys.path.insert(0, os.getenv("BUILDTEST_ROOT"))
+sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
Why is this line needed?

shahzebsiddiqui replied:
when i remove the line i run into import error

```
ssi29@lsfportal1> python examples/moduletest.py
Traceback (most recent call last):
  File "examples/moduletest.py", line 6, in <module>
    from buildtest.modules.util import Module
ModuleNotFoundError: No module named 'buildtest.modules'
```

Pretty sure its the packaging or something else i couldnt figure out
```
ssi29@lsfportal1> pip show -f buildtest-framework
Name: buildtest-framework
Version: 0.7.7
Summary: HPC Application Testing Framework
Home-page: https://github.com/HPC-buildtest/buildtest-framework
Author: Shahzeb Siddiqui
Author-email: shahzebmsiddiqui@gmail.com
License: GPLv2
Location: /u/users/ssi29/.local/lib/python3.7/site-packages/buildtest_framework-0.7.7-py3.7.egg
Requires: argcomplete, PyYAML, distro, termcolor
Required-by:
Files:
Cannot locate installed-files.txt
```

Looking into my egg i dont have the updated files such as ``buildtest.modules``

```
ssi29@lsfportal1> ll /u/users/ssi29/.local/lib/python3.7/site-packages/buildtest_framework-0.7.7-py3.7.egg/buildtest/
total 24
drwxr-xr-x 3 ssi29 hks 4096 Feb 22 17:13 docgen
-rw-r--r-- 1 ssi29 hks   60 Feb 22 17:13 __init__.py
-rw-r--r-- 1 ssi29 hks  739 Feb 22 17:13 main.py
drwxr-xr-x 2 ssi29 hks 4096 Feb 22 17:13 __pycache__
drwxr-xr-x 2 ssi29 hks 4096 Feb 22 17:13 scripts
-rw-r--r-- 1 ssi29 hks 1964 Feb 22 17:13 settings.yml
drwxr-xr-x 6 ssi29 hks 4096 Feb 22 17:13 tools

```

Hence i added that line so that it works with my git repo. I know i am missing something in my environment

vsoch replied:
Did you install it?

vsoch replied:
Yeah I don't see modules - I don't think it's installed. Is there an __init__.py in the folder?

shahzebsiddiqui committed:
adding Spider() class which allows user to programmatically query
any module tree. This works well with Module class where you can
query all modules from Spider() and test them with Module() class

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+        #  if user specifies a tree, then run spider command otherwise read from configuration file.
+        if tree:
+            self.tree = tree
+            spider_cmd = f"{os.getenv('LMOD_DIR')}/spider -o spider-json {self.tree}"
```
If this isn't defined, it's going to return as None, and then the command will be `None/spider` and should not be run. We need to have a check here that the LMOD_DIR envar is defined, and that it's a path that is valid before executing anything to the system.

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+            self.tree = tree
+            spider_cmd = f"{os.getenv('LMOD_DIR')}/spider -o spider-json {self.tree}"
+
+            out = subprocess.check_output(spider_cmd, shell=True).decode("utf-8")
```
Keeping in mind we are going to refactor BuidlTestCommand, why not use it here? And then (in the future) when we refactor, we can remove the shell=True.

shahzebsiddiqui replied:
I did try using `BuildTestCommand` i had trouble getting content into a dictionary using `json.loads()` when using `get_output()` method so i gave up and used `subprocess` for now. I know this needed to be fixed.

vsoch replied:
okay I'll add a note about this in the issue for it.

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+            spider_cmd = f"{os.getenv('LMOD_DIR')}/spider -o spider-json {self.tree}"
+
+            out = subprocess.check_output(spider_cmd, shell=True).decode("utf-8")
+            self.spider_content = json.loads(out)
```
I could see issues with loading this here - it's not imminent to put a try except, but let's keep it in mind.

shahzebsiddiqui replied:
Let's fix it for next time. It will show up during regtest and coverage reports

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+            self.spider_content = json.loads(out)
+
+        else:
+            with open(BUILDTEST_SPIDER_FILE, "r") as fd:
```
This assumes that the file exists, which I don't think we can safely do.

shahzebsiddiqui replied:
i was not sure about the ``else`` logic. We could simply use `MODULEPATH` instead of this file.

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+            with open(BUILDTEST_SPIDER_FILE, "r") as fd:
+                self.spider_content = json.load(fd)
+
+            content = load_configuration(BUILDTEST_CONFIG_FILE)
```
Is this guaranteed to load something, and is the something guaranteed to have `BUILDTEST_MODULEPATH`? The load_configuration should likely exit cleanly if anything weird happens and then the self.tree should get like:

```
self.tree = content.get("BUILDTEST_MODULEPATH", [])
```

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+        :rtype: list
+        """
+
+        return sorted(list(self.spider_content.keys()))
```
You don't need the list here, the sorted keys already is a list. Here is an example:

```
test = {1:1, 2:2, 3:3}
sorted(test)                                                                                                            
[1, 2, 3]

type(sorted(test))                                                                                                      
list
```

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+        """Retrieve all module names from all module tree.
+
+        :return: returns  a sorted list of all full canonical module name from all spider records.
+        :rtype: list
```
Now that I see this - should this (and or the get_unique_software) be an iterator? How / what is the context in which this is used?

shahzebsiddiqui replied:
This is used in context of automating ``module load test``. 

Also for some reason the list is not sorted when testing the output

```
Unique Software: ['gompi', 'numactl', 'GCCcore', 'flex', 'ncurses', 'OpenMPI', 'Anaconda3', 'bzip2', 'foss', 'gettext', 'GCC', 'libiconv', 'libtool', 'libffi', 'GMP', 'Tcl', 'libpciaccess', 'Python', 'xz', 'gdbm', 'lmod', 'Automake', 'util-macros', 'Autotools', 'EasyBuild', 'libxml2', 'XZ', 'binutils', 'SQLite', 'tar', 'xorg-macros', 'OSU-Micro-Benchmarks', 'Autoconf', 'settarg', 'libreadline', 'OpenBLAS', 'ScaLAPACK', 'CUDA', 'CMake', 'FFTW', 'Clang', 'PyCharm', 'hwloc', 'Bison', 'm4', 'pkgconf', 'zlib', 'M4', 'libsigsegv', 'readline', 'help2man', 'diffutils']

```

so ``list(self.spider_content.keys()`` is not sorted.

vsoch commented on buildtest/modules/spider.py:
```diff
@@ -0,0 +1,103 @@
+                    for parent_comb in self.spider_content[module][mpath]["parentAA"]:
+                        [parent_set.add(parent_module) for parent_module in parent_comb]
+
+        return sorted(list(parent_set))
```
Another sorted ( list , you can remove the list I think.

vsoch commented on examples/spider_example.py:
```diff
@@ -0,0 +1,36 @@
+import os
+import sys
+
+sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
We shouldn't need this if buildtest has been installed.

vsoch commented on examples/spider_example.py:
```diff
@@ -0,0 +1,36 @@
+
+from buildtest.modules.spider import Spider
+from buildtest.modules.util import Module
+
```
An organizational question - why is Module imported from util? It would be more intuitive to have like:

```
from buildtest.modules import Module
```
But you don't want it accessible via the init?

shahzebsiddiqui replied:
this makes sense to me if the file is `buildtest/modules.py` i can do that. I was not sure if i will be implementing other class so for instance `buildtest/modules/spider.py` was added to avoid having one massive file with all classes and methods.

I can make the file change if it makes sense.

vsoch replied:
But you don't want to write it into an __init__.py in the modules folder?

shahzebsiddiqui committed:
move files up a directory from buildtest/modules/ and update example
files based on comments by @vsoch.
